databuilder/sqlalchemy_utils.py

```diff
@@ -0,0 +1,207 @@
+
+    This means that all the setup/cleanup logic for temporary tables can be bundled up
+    with the queries that use them and they no longer need to be manged out-of-band.
+    """
+
+    setup_queries = ()
+    cleanup_queries = ()
+
+
+def get_setup_and_cleanup_queries(clause):
+    """
+    Given a SQLAlchemy ClauseElement find all TemporaryTables embeded in it and return a
+    pair of:
+
+        setup_queries, cleanup_queries
+
+    which are the combination of all the setup and cleanup queries from those
+    TemporaryTables in the correct order for execution.
+
+    There's obviously a bit of algorithmic complexity here, but it's a fairly generic,
+    graph traversal kind of complexity, contained in one place, which allows us to avoid
+    structural complexity elsewhere in the code so I think it's worth it.
+    """
+    # TemporaryTables can be arbitrarily nested in that their setup queries can
+    # themselves contain references to TemporaryTables and so on. So we need to
+    # recursively unpack these, taking note of a number of subtleties:
+    #
+    #  1. We need to keep track of what level of nesting each table is at. The setup
+    #     queries for the most deeply nested tables must be run first so they exist by
+    #     the time queries for their dependant tables get run.
+    #
+    #  2. The same table can occur at multiple levels of nesting so we need to track the
+    #     highest level for each table.
+    #
+    #  3. By construction there can be no cycles in this graph *except* that each
+    #     table's setup queries will contain a reference to itself (as part of a "create
+    #     this table" statement). This means we need to keep track of which tables we've
+    #     seen on each given branch to avoid looping.
+    clauses = [(clause, 0, set())]
+    table_levels = {}
+    while clauses:
+        next_clause, level, seen = clauses.pop(0)
+        for table in get_temporary_tables(next_clause):
+            if table in seen:
+                continue
+            # We may end up setting this multiple times, but if so we'll set the highest
+            # level last, which is what we want
+            table_levels[table] = level
+            # Add all of the tables setup and cleanup queries to the stack to be checked
+            clauses.extend(
+                [
+                    (query, level + 1, seen | {table})
+                    for query in table.setup_queries + table.cleanup_queries
+                ]
+            )
+
+    # Sort tables in reverse order by level
+    tables = sorted(
+        table_levels.keys(),
+        key=table_levels.__getitem__,
+        reverse=True,
+    )
+
+    setup_queries = []
+    cleanup_queries = []
+    for table in tables:
+        setup_queries.extend(table.setup_queries)
+        # Add cleanup queries to the start of the list so they end up in mirror image
+        # order to the setup queries
+        cleanup_queries[:0] = table.cleanup_queries
```

Comment:
That wouldn't quite give the behaviour we want though (if I'm understanding correctly) which is to keep each table's cleanup queries in their original order, but have the table order reversed. So if we have:
```
table_1: [a, b, c], [x, y, z]
table_2: [1, 2, 3], [4, 5, 6]
```

We want the combined queries to be:
```
[a, b, c, 1, 2, 3], [4, 5, 6, x, y, z]
```

Whereas extending and reversing would give us:
```
[a, b, c, 1, 2, 3], [6, 5, 4, z, y, x]
```

But thinking about it more, it might be clearer to iterate over the tables twice: once forward to build the setup queries, and once backwards to build the cleanup queries.